File: arvores/java/generic/src/generictree/GenericTree.java
package generictree;

import java.util.ArrayList;
import java.util.List;

public class GenericTree<T> {
	private Node<T> root;
	private int size;

	public GenericTree() {
		root = null;
		size = 0;
	}

	public int size() {
		return size;
	}

	public boolean isEmpty() {
		return size == 0;
	}

	public List<T> elements() {
		List<T> elements = new ArrayList<>();
		collectElements(elements, root);
		return elements;
	}

	private void collectElements(List<T> list, Node<T> node) {
		list.add(node.element());
		for (Node<T> child : node.getChildren()) {
			collectElements(list, child);
		}
	}

	public List<Position<T>> positions() {
		List<Position<T>> positions = new ArrayList<>();
		collectPositions(positions, root);
		return positions;
	}

	private void collectPositions(List<Position<T>> list, Node<T> node) {
		list.add(node);
		for (Node<T> child : node.getChildren()) {
			collectPositions(list, child);
		}
	}

	public void replace(Position<T> position, T element) {
		Node<T> node = validate(position);
		node.setElement(element);
	}

	public Position<T> root() {
		return root;
	}

	public Position<T> parent(Position<T> position) {
		Node<T> node = validate(position);
		return node.getParent();
	}

	public List<Position<T>> children(Position<T> position) {
		Node<T> node = validate(position);
		return new ArrayList<>(node.getChildren());
	}

	public boolean isExternal(Position<T> position) {
		Node<T> node = validate(position);
		return node.getChildren().isEmpty();
	}

	public boolean isRoot(Position<T> position) {
		Node<T> node = validate(position);
		return node == root;
	}

	private Node<T> validate(Position<T> position) {
		if (!(position instanceof Node)) {
			throw new IllegalArgumentException("Invalid position type");
		}
		Node<T> node = (Node<T>) position;
		if (node.getParent() == node) { // Convention to indicate deleted
			throw new IllegalArgumentException("Position is no longer in the tree");
		}
		return node;
	}

	public Position<T> add(T element, Position<T> parent) {
		if (!isEmpty() && parent == null) {
			throw new IllegalArgumentException("Parent position can't be null for a non empty tree");
		}
		Node<T> parentNode = parent == null ? null : validate(parent);
		Node<T> newNode = new Node<>(element, parentNode);
		if (parentNode == null) {
			root = newNode;
		} else {
			parentNode.addChild(newNode);
		}
		size++;
		return newNode;
	}

	public void remove(Position<T> position) {
		Node<T> node = validate(position);
		if (node == root) {
			root = null;
		} else {
			Node<T> parent = node.getParent();
			if (parent != null) {
				parent.removeChild(node);
			}
		}
		size -= subtreeSize(node);
		markAsRemoved(node);
	}

	private int subtreeSize(Node<T> node) {
		int subtreeCount = 1;
		for (Node<T> child : node.getChildren()) {
			subtreeCount += subtreeSize(child);
		}
		return subtreeCount;
	}

	private void markAsRemoved(Node<T> node) {
		node.setParent(node);
		for (Node<T> child : node.getChildren()) {
			markAsRemoved(child);
		}
	}

	public Position<T> find(T element) {
		return findRecursive(root, element);
	}

	private Position<T> findRecursive(Node<T> node, T target) {
		if (node == null) {
			return null;
		}
		if (node.element().equals(target)) {
			return node;
		}
		for (Node<T> child : node.getChildren()) {
			Position<T> found = findRecursive(child, target);
			if (found != null) {
				return found;
			}
		}
		return null;
	}
}

File: arvores/csharp/generic/GenericTree.cs
namespace GenericTree
{
    public class GenericTree<T>
    {
        private Node? root;
        private int size;

        public GenericTree()
        {
            root = null;
            size = 0;
        }

        public int Size() => size;

        public bool IsEmpty() => size == 0;

        public List<T> Elements()
        {
            List<T> elements = new List<T>();
            CollectElements(elements, root);
            return elements;
        }

        private void CollectElements(List<T> list, Node? node)
        {
            if (node != null)
            {
                list.Add(node.Element());
                foreach (Node child in node.GetChildren())
                {
                    CollectElements(list, child);
                }
            }
        }

        public List<IPosition<T>> Positions()
        {
            List<IPosition<T>> positions = new List<IPosition<T>>();
            CollectPositions(positions, root);
            return positions;
        }

        private void CollectPositions(List<IPosition<T>> list, Node? node)
        {
            if (node != null)
            {
                list.Add(node);
                foreach (Node child in node.GetChildren())
                {
                    CollectPositions(list, child);
                }
            }
        }

        public void Replace(IPosition<T>? position, T element)
        {
            Node node = Validate(position);
            node.SetElement(element);
        }

        public IPosition<T>? Root()
        {
            return root;
        }

        public IPosition<T>? Parent(IPosition<T> position)
        {
            Node node = Validate(position);
            return node.GetParent();
        }

        public List<IPosition<T>> Children(IPosition<T> position)
        {
            Node node = Validate(position);
            return new List<IPosition<T>>(node.GetChildren());
        }

        public bool IsExternal(IPosition<T> position)
        {
            Node node = Validate(position);
            return node.GetChildren().Count == 0;
        }

        public bool IsRoot(IPosition<T> position)
        {
            Node node = Validate(position);
            return node == root;
        }

        private Node Validate(IPosition<T>? position)
        {
            if (!(position is Node node))
                throw new ArgumentException("Invalid position type");

            if (node.GetParent() == node) // Convention to indicate deleted
                throw new ArgumentException("Position is no longer in the tree");

            return node;
        }

        public IPosition<T> Add(T element, IPosition<T>? parent)
        {
            if (!IsEmpty() && parent == null)
            {
                throw new ArgumentException("Parent position can't be null for a non empty tree");
            }
            Node? parentNode = parent == null ? null : Validate(parent);
            Node newNode = new Node(element, parentNode);
            if (parentNode == null)
            {
                root = newNode;
            }
            else
            {
                parentNode.AddChild(newNode);
            }
            size++;
            return newNode;
        }

        public void Remove(IPosition<T> position)
        {
            Node node = Validate(position);
            if (node == root)
            {
                root = null;
            }
            else
            {
                Node? parent = node.GetParent();
                if (parent != null)
                {
                    parent.RemoveChild(node);
                }
            }
            size -= SubtreeSize(node);
            MarkAsRemoved(node);
        }

        private int SubtreeSize(Node node)
        {
            int subtreeCount = 0;
            foreach (Node child in node.GetChildren())
            {
                subtreeCount += SubtreeSize(child);
            }
            return 1 + subtreeCount;
        }

        private void MarkAsRemoved(Node node)
        {
            node.SetParent(node);
            foreach (Node child in node.GetChildren())
            {
                MarkAsRemoved(child);
            }
        }

        public IPosition<T>? Find(T element)
        {
            return FindRecursive(root, element);
        }

        private IPosition<T>? FindRecursive(Node? node, T target)
        {
            if (node == null)
            {
                return null;
            }
            if (target != null && target.Equals(node.Element()))
            {
                return node;
            }
            foreach (Node child in node.GetChildren())
            {
                IPosition<T>? found = FindRecursive(child, target);
                if (found != null)
                {
                    return found;
                }
            }
            return null;
        }

        // Inner class Node
        private class Node : IPosition<T>
        {
            private T element;
            private Node? parent;
            private List<Node> children;

            public Node(T element, Node? parent)
            {
                this.element = element;
                this.parent = parent;
                children = new List<Node>();
            }

            public T Element() => element;

            public void SetElement(T element) => this.element = element;

            public Node? GetParent() => parent;

            public void SetParent(Node? parent) => this.parent = parent;

            public List<Node> GetChildren() => children;

            public void AddChild(Node child)
            {
                children.Add(child);
                child.SetParent(this);
            }

            public void RemoveChild(Node child)
            {
                children.Remove(child);
                child.SetParent(null);
            }
        }
    }
}
